examples/local/vstream_client.go

```diff
@@ -82,17 +72,372 @@ func main() {
+	// ********************************************************************************************************
+	var lastEventReceivedAtUnix atomic.Int64
+
+	ctx, heartbeatCancel := context.WithCancel(ctx)
+	heartbeat := time.NewTicker(heartbeatDur)
+	defer heartbeat.Stop()
+
+	go func() {
+		for {
+			select {
+			case tm := <-heartbeat.C:
+				// if we haven't received an event yet, we'll skip the heartbeat check
+				if lastEventReceivedAtUnix.Load() == 0 {
+					continue
+				}
+
+				// if we haven't received an event in twice the heartbeat duration, we'll cancel the context, since
+				// we're likely disconnected, and exit the goroutine
+				if tm.Sub(time.Unix(lastEventReceivedAtUnix.Load(), 0)) > heartbeatDur*2 {
+					heartbeatCancel()
+					return
+				}
+
+			case <-ctx.Done():
+				// since heartbeatCtx is derived from ctx, there's no need to cancel it
+				return
+			}
+		}
+	}()
+
+	// ********************************************************************************************************
+	// Flush Data + Store Vgtid
+	//
+	// as we process events, we'll periodically need to check point the last vgtid and store the customers in the
+	// database. You can control the frequency of this flush by adjusting the minFlushDuration and maxCustomersToFlush.
+	// This is only called when we have an event that guarantees we're not flushing mid-transaction.
+	// ********************************************************************************************************
+	var lastFlushedVgtid, latestVgtid *binlogdatapb.VGtid
+	var lastFlushedAt time.Time
+
+	flushFunc := func() error {
+		// if the lastFlushedVgtid is the same as the latestVgtid, we don't need to do anything
+		if proto.Equal(lastFlushedVgtid, latestVgtid) {
+			return nil
+		}
+
+		// if it hasn't been long enough since the last flush, and we haven't exceeded our match batch size, don't
+		// flush. We can always replay as needed.
+		if time.Since(lastFlushedAt) < minFlushDuration && len(customers) < maxCustomersToFlush {
+			return nil
+		}
+
+		// if the customer db is the same db you're storing the vgtid in, you could do both in the same transaction
+
+		// flush the customers to the database, using the max batch size
+		for customerChunk := range slices.Chunk(customers, maxCustomersToFlush) {
+			err := upsertCustomersToDB(ctx, customerChunk)
+			if err != nil {
+				return err
+			}
+		}
+
+		// reset the customers slice to free up memory. If you really want to be efficient, you could reuse the slice.
+		customers = slices.Delete(customers, 0, len(customers))
+
+		// always store the latest vgtid, even if there are no customers to store
+		err := storeLastVgtid(ctx, latestVgtid)
+		if err != nil {
+			return err
+		}
+
+		lastFlushedVgtid = latestVgtid
+		lastFlushedAt = time.Now()
+
+		return nil
+	}
+
+	// ********************************************************************************************************
```

Comment:
I tried to add comments to each case, but it's limited by my understanding, so please feel free to correct anything that isn't right. I think it would make sense to transfer a lot of the comments over to the proto, which has some very useful comments, but most of the type enums have none.